Add create event endpoint

reupen opened the pull request:
Still a few things to clarify, so may be a few changes next week.

Decided to leave programmes in the event app for now, for review later on if it becomes a generic concept. Have still included them in the loadmetadata command as I don't think we need another command. Could perhaps add some arguments to it though.

In relation to the null=False/required=True thing, a different workaround that seems to work is specifying `default=''` on the model field. Don't have much preference really.

marcofucci commented on datahub/event/models.py:
```diff
@@ -0,0 +1,59 @@
+    start_date = models.DateField(null=True, blank=True)
+    end_date = models.DateField(null=True, blank=True)
+    location_type = models.CharField(
+        max_length=MAX_LENGTH, null=True, blank=True, choices=LOCATION_TYPES
```
Can't we remove `null=True`?

reupen replied:
Yep, I'll get rid of it.

marcofucci commented on datahub/event/models.py:
```diff
@@ -0,0 +1,59 @@
+    lead_team = models.ForeignKey(
+        'metadata.Team', on_delete=models.PROTECT, null=True, blank=True, related_name='+'
+    )
+    additional_teams = models.ManyToManyField('metadata.Team', blank=True, related_name='+')
```
wouldn't it be better to have a through model with a lead flag? What if you want to change the lead team? You would have to add the old `lead_team` to the `additional_teams`, remove the new lead from the `additional_teams` and set it to the `lead_team`.

reupen replied:
It reflects the API (which in turn reflects the UI). But in the other place I used through (investment project) it brought about a lot of complexity. It could equally be fixed by changing additional_teams to teams and then you only need to validate that lead_team is in teams.

Anyway I think this equally applies to the API so might be better discussed there: https://github.com/uktrade/data-hub-api-spec/pull/35

elcct replied:
Or we can have `teams` that contain all teams and `lead_team` as a "pointer" to the team that is leading. That removes complexity @marcofucci mentioned.

elcct replied:
@reupen haha commented same time and looks like we think same thing :)

marcofucci replied:
👍  for `teams`

elcct commented on datahub/event/models.py:
```diff
@@ -0,0 +1,59 @@
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4)
+    name = models.CharField(max_length=MAX_LENGTH)
+    event_type = models.CharField(max_length=MAX_LENGTH, choices=EVENT_TYPES)
+    start_date = models.DateField(null=True, blank=True)
```
Are we going to have event without start date?

reupen replied:
At the moment on JIRA it says it is optional. It might also need to be nullable for the migration.

marcofucci replied:
Yeah it doesn't seem to make sense. We should find out why it's nullable and fix the data during the migration if that is the case.

reupen commented:
@marcofucci @elcct Have made some changes (changed additional teams to teams, added organiser, changed event and location type to be foreign keys).

There might be some other changes needed, but I'll do those in separate PRs.

marcofucci commented on datahub/event/serializers.py:
```diff
@@ -0,0 +1,64 @@
+class EventSerializer(serializers.ModelSerializer):
+    """Event serialiser."""
+
+    _lead_team_error_text = 'Lead team must be in teams array.'
```
You could try use the same DRF logic.
Define the message like this: 
```
default_error_messages = {
    'lead_in_teams': _('Lead team must be in teams array.'),
}
```
and then use it this way:
```self.error_messages['lead_in_teams']```

It might work ;)

reupen replied:
Done!

marcofucci commented on datahub/event/constants.py:
```diff
@@ -0,0 +1,22 @@
+class EventType(Enum):
+    """Event type constants."""
+
+    seminar = Constant('Seminar', '771f654d-e6b1-4fad-8a89-b6ac44147830')
```
I'm trying to stay away from creating more constants and just get a random record in the tests in order to avoid duplications.
I know that this is more performant though so I'm easy on this, we could refactor later on.

reupen replied:
I don't like them either, but we don't have another solution at present.

marcofucci replied:
Sorry I wasn't clear, the other solution is to use `EventType.objects.first()` instead.
It's less performant so I'll leave it to you to decide.

reupen commented on datahub/core/management/commands/loadmetadata.py:
```diff
@@ -4,7 +4,29 @@
+    EVENTS_FIXTURE_DIR / 'programmes.yaml',
+    EVENTS_FIXTURE_DIR / 'location_types.yaml',
+    EVENTS_FIXTURE_DIR / 'event_types.yaml',
+)
```
Yeah, that might be better.

marcofucci commented on datahub/event/serializers.py:
```diff
@@ -0,0 +1,67 @@
+            'address_county',
+            'address_postcode',
+            'address_town',
+            'end_date',
```
Do you need to check if `end_date <= start_date` ?

reupen replied:
Yep, good point. I'll do it in another PR if that's OK.

marcofucci commented on datahub/event/test/test_views.py:
```diff
@@ -0,0 +1,179 @@
+                'id': Team.crm.value.id,
+                'name': Team.crm.value.name,
+            },
+            'teams': [{
```
asserting lists fails in my local env 🤔

reupen replied:
Ah, sorry, the order would be undefined. Another PR coming up!
